src/main/java/io/vertx/core/shareddata/impl/LocalAsyncLocks.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2011-2018 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+
+package io.vertx.core.shareddata.impl;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Context;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.VertxException;
+import io.vertx.core.shareddata.Lock;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Stream;
+
+import static java.util.stream.Collectors.*;
+
+/**
+ * @author Thomas Segismont
+ */
+public class LocalAsyncLocks {
+
+  // Immutable
+  private static class WaitersList {
+
+    final List<LockWaiter> waiters;
+
+    WaitersList(List<LockWaiter> waiters) {
+      this.waiters = waiters;
+    }
+
+    int size() {
+      return waiters.size();
+    }
+
+    LockWaiter first() {
+      return waiters.get(0);
+    }
+
+    WaitersList add(LockWaiter waiter) {
+      return new WaitersList(Stream.concat(waiters.stream(), Stream.of(waiter)).collect(toList()));
+    }
+
+    WaitersList removeStale() {
```

Comment:
remove this `removeStale`

it adds a cleanup mechanism that is already taken care of by acquireLock() already

it makes the algorithm simpler and smaller

Reply:
I don't understand. `acquireLock()` invokes this method to remove the first waiter and the following ones if they timed out.

Reply:
I mean it could be simpler

- since `acquireLock ` already takes care of skipping the lock if that's timed out, each lock will be either executed or skipped, so there is no need to cleanup the stale elements
- when `compute()` is called on the map, it returns the rest of the list or `null` if the list was a single element

I don't think it buys much to filter stale locks

Reply:
Ha. I see your point now. Filtering out eagerly the waiters removes the need to recompute the waiters list N times if there are N timed-out waiters.

Reply:
I just pushed a commit:

* removed WaitersList abstraction
* do not skip timed out waiters when recomputing the list